Run downsampling using persistent tasks

salvatore-campagna opened the pull request:
Add support for downsampling persistent tasks. We would like to have the ability
to resume downsampling tasks, instead of starting from scratch in case of 
("retriable") failures.

Instead of keeping track of the state of the task we just try to query the downsampling
target index before starting the actual downsampling task, getting the latest document
indexed and its `tsid`. If we find any, we just restart from that `tsid`, maybe
overwriting a subset of the documents already indexed (documents with the same
tsid whose timestamp is smaller then the one of the latest document).

Querying the downsampling target index is possible after introducing a predictable
naming scheme for the target index. This was not the case before, a part of the target
index name was random. That is required to be able to start from where the task left.

Note that since the ordering of documents is based on (tsid, timestamp) we do not
include the timestamp in the query to avoid skipping downsampling of documents
with a larger tsid but smaller timestamp.

Note also that we are not removing the `TransportDownsampleIndexerAction` for 
backward compatibility. An older master node running a previous version would
not use persistent tasks and an API call would result in something like "missing
downsample action".

We are also adding an additional downsample rest query parameter, `timeout`.
It allows a user to set the maximum  time taken while waiting for a downsampling
task to complete before returning with an timeout error. Note that when the timeout
triggers the task will still run under control of the executor, just waiting for it to finish
will result in a timeout. The default value for this timeout is 1 day.

Resolve #93582

salvatore-campagna committed:
feature: run downsampling using persistent tasks

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @salvatore-campagna, I've created a changelog YAML for you.

salvatore-campagna committed:
Update docs/changelog/97557.yaml

salvatore-campagna committed:
Update docs/changelog/97557.yaml

elasticsearchmachine commented:
Hi @salvatore-campagna, I've updated the changelog YAML for you.

salvatore-campagna committed:
fix: npe and exceptions

salvatore-campagna commented:
I removed `x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/TransportDownsampleIndexerAction.java` but I will restore it for BWC.

salvatore-campagna commented:
Will get rid of all `deleteIndex`.

salvatore-campagna committed:
fix: mark persistent task completion

salvatore-campagna committed:
fix: resume task from latest written tsid

salvatore-campagna committed:
fix: do not delete the downsampling target index

salvatore-campagna committed:
fix: handle existence of downsampling task

salvatore-campagna committed:
fix: introduce downsampling request timeout

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardPersistentTaskExecutor.java:
```diff
@@ -0,0 +1,128 @@
+        // Here we make sure we assign the task to the actual node holding the shard identified by
+        // the downsampling task shard id.
+        final ShardId shardId = params.shardId();
+        final ShardRouting shardRouting = clusterState.routingTable().shardRoutingTable(shardId).shard(shardId.id());
```
I think we also  need to check whether the shard routing is started here.

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java:
```diff
@@ -182,11 +211,44 @@ public DownsampleIndexerAction.ShardDownsampleResponse execute() throws IOExcept
+                null,
+                Collections.emptyMap()
+            );
+            return new BoolQueryBuilder().filter(new TermQueryBuilder(TimeSeriesIdFieldMapper.NAME, this.state.tsid()))
```
I think we can use Lucene query api instead of Elasticsearch's `QueryBuilder` api. That way there is no need to create a `SearchExecutionContext` here. For example: 

```
return new BooleanQuery.Builder().add(new TermQuery(new Term(TimeSeriesIdFieldMapper.NAME, this.state.tsid())), BooleanClause.Occur.FILTER).build();
```

martijnvg replied:
Actually this query will only match with documents that only have the specified tsid. I think we want to match with all documents that have the specified tsid and beyond? So I think we need a range query? If that is the case we need `SortedSetDocValuesField.newSlowRangeQuery(...)` because the _tsid field isn't indexed.

jimczi replied:
Hey, just a heads up, we've got this speedy query built for sorted indices that could totally work here too:
[link](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/lucene/queries/SearchAfterSortedDocQuery.java). 
Adapting it to our needs shouldn't be a big deal. And if we find it handy, we might even think about adding it directly into Lucene, kind of like what's been done with IndexSortSortedSetDocValuesRangeQuery.

martijnvg replied:
Thanks for pointing this out Jim. The main focus is to get this PR merged sooner than later. But we will definitely change this in a followup change.

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java:
```diff
@@ -182,11 +211,44 @@ public DownsampleIndexerAction.ShardDownsampleResponse execute() throws IOExcept
+            );
+            return new BoolQueryBuilder().filter(new TermQueryBuilder(TimeSeriesIdFieldMapper.NAME, this.state.tsid()))
+                .toQuery(searchExecutionContext);
+        } else if (this.state.done()) {
```
maybe return in the `execute()` instead of here? This we can check without running query.

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java:
```diff
@@ -157,7 +174,13 @@ public DownsampleIndexerAction.ShardDownsampleResponse execute() throws IOExcept
 
         if (task.getNumIndexed() != task.getNumSent()) {
-            task.setRollupShardIndexerStatus(RollupShardIndexerStatus.FAILED);
+            task.updatePersistentTaskState(
```
maybe we should update the `task.updatePersistentTaskState(...)` in `RollupShardPersistentTaskExecutor#nodeOperation(...)`? When can just catch the exceptions thrown here there and set the status to failed.

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollupShardIndexerStatus.java:
```diff
@@ -7,13 +7,51 @@
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        out.writeVInt(this.ordinal);
```
Did we before serialise this enum as string?

salvatore-campagna replied:
I will check..in that case I will revert it...

salvatore-campagna replied:
This was not a Writable before.

martijnvg reviewed (commented):
I did a first review round. Thanks for working on this!

I think we should write a java integration test, that tests performing downsampling while all nodes get restarted. A similar test exists for CCR, see: `FollowerFailOverIT`.

salvatore-campagna committed:
fix: wait for existing task

salvatore-campagna committed:
fix: checkstyle code format violation

salvatore-campagna committed:
fix: just one one byte to serialize the enum

salvatore-campagna committed:
fix: serielization of state and params

salvatore-campagna committed:
fix: include registry in test

salvatore-campagna committed:
fix: compilation error

salvatore-campagna committed:
fix: compilation error

salvatore-campagna committed:
fix: downsampling tests

salvatore-campagna committed:
fix: remove blocking calls from cluster updates

salvatore-campagna committed:
fix: some serialization issues

salvatore-campagna committed:
fix: catch possible query failure and start task from scratch

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardPersistentTaskExecutor.java:
```diff
@@ -0,0 +1,147 @@
+        final SearchHit[] lastRollupTsidHits = client.prepareSearch(params.rollupIndex())
+            .addSort(TimeSeriesIdFieldMapper.NAME, SortOrder.DESC)
+            .setSize(1)
+            .setQuery(new MatchAllQueryBuilder()) // TODO: can we do better then querying all documents?
```
I think we need to set search preference to `_shard: shardID`, otherwise we query all shards or the target index. And we get tsids that this `RollupShardIndexer` never downsampled. This should work because the target index always has the same number of shards of the source index. So tsids in source index shard 2 will be indexed in target index shard 2.

salvatore-campagna committed:
fix: shard id from string deserialization

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardPersistentTaskExecutor.java:
```diff
@@ -0,0 +1,154 @@
+    @Override
+    protected void nodeOperation(final AllocatedPersistentTask task, final RollupShardTaskParams params, final PersistentTaskState state) {
+        // NOTE: query the downsampling target index so that we can start the downsampling task from the latest indexed tsid.
+        SearchHit[] lastRollupTsidHits;
```
I think we can just use the search method that accepts an action listener here? I the onResponse we just need to fork back to a downsample thread.

salvatore-campagna replied:
It looks like using client.search I do not have the option to specify things like sorting.

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardPersistentTaskExecutor.java:
```diff
@@ -0,0 +1,154 @@
+        // Here we make sure we assign the task to the actual node holding the shard identified by
+        // the downsampling task shard id.
+        final ShardId shardId = params.shardId();
+        final ShardRouting shardRouting = clusterState.routingTable().shardRoutingTable(shardId).primaryShard();
```
I think any shard is ok for downsampling? Also replica shards.

salvatore-campagna replied:
Well in the transport action we set replicas to 0 and we set the original value back for the index only after downsampling is complete.  Here I used only primary shards because I wanted to make sure we are not using some replica that maybe is created before we set the number of replicas to 0.

salvatore-campagna committed:
test: exception while writing read only index

salvatore-campagna committed:
fix: add back TransportDownsampleIndexerAction.class for bwc

salvatore-campagna committed:
Merge branch 'main' into feature/93582-downsample-persistent-task

salvatore-campagna committed:
fix: remove unused SearchExecutionContext

salvatore-campagna committed:
fix: mark task failure in case of exception

salvatore-campagna committed:
fix: serielize enum as a string for bwc

salvatore-campagna committed:
fix: set the shard id preference

salvatore-campagna committed:
fix: include missing timeout

salvatore-campagna committed:
fix: remove unused throws

salvatore-campagna committed:
fix: search last tsid using custom thread pool

salvatore-campagna committed:
fix: make execute return void

salvatore-campagna committed:
fix: simplify status update and restore executet return value

salvatore-campagna committed:
run the indexer even if searching throws

salvatore-campagna committed:
fix: shard id serialization

salvatore-campagna committed:
fix: shard id serialization equals and hashCode

salvatore-campagna committed:
fix: ShardIn construction from StreamInput

salvatore-campagna committed:
fix: string array serialization using lists

salvatore-campagna committed:
fix: use an ObjectParser for both state and params

salvatore-campagna commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardPersistentTaskExecutor.java:
```diff
@@ -0,0 +1,161 @@
+        );
+        try {
+            rollupShardIndexer.execute();
+            task.markAsCompleted();
```
This is a duplicate...will get rid of it.

salvatore-campagna committed:
Merge branch 'main' into feature/93582-downsample-persistent-task

salvatore-campagna committed:
fix: a few serialization issues

salvatore-campagna committed:
fix: some seriealization and synchronization issues

salvatore-campagna committed:
fix: timeout is optional

salvatore-campagna committed:
fix: add missing timeout parameter

salvatore-campagna committed:
fix: use predictable index name

salvatore-campagna committed:
fix: expand wildcard

salvatore-campagna committed:
log: improve logging

salvatore-campagna committed:
test: mvoe to internalClusterTest

salvatore-campagna committed:
fix: remove test after removing TransportDownsampleIndexerAction

salvatore-campagna committed:
test: duplicate downsample request on same index with same params

salvatore-campagna committed:
fix: use xor

salvatore-campagna committed:
Merge branch 'main' into feature/93582-downsample-persistent-task

salvatore-campagna committed:
fix: version id

martijnvg commented on x-pack/plugin/rollup/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/DownsampleTransportFailureIT.java:
```diff
@@ -339,40 +354,4 @@ public void testDownsampleActionExceptionDisruption() {
         assertDownsampleFailure(testCluster.coordinatorName());
     }
-
-    public void testRollupIndexerActionExceptionDisruption() {
```
Maybe also add a test with multiple nodes and performs a rolling restart while performing downsampling?
Doesn't have to be in this class, but can also be in a new test class.

salvatore-campagna replied:
I wrote tests doing full and rolling restarts. Consider that, anyway, if the indexing shard (where we write downsample data) is not ready when we send the bulk request the task will incur into a failure.

martijnvg commented on server/src/main/java/org/elasticsearch/action/downsample/DownsampleConfig.java:
```diff
@@ -55,19 +58,23 @@ public class DownsampleConfig implements NamedWriteable, ToXContentObject {
     private final DateHistogramInterval fixedInterval;
     private final String timeZone = DEFAULT_TIMEZONE;
     private final String intervalType = FIXED_INTERVAL;
+    private final TimeValue timeout;
```
Should this perhaps be a cluster setting instead? I what case would someone want to configure a different timeout for different downsample operations? Or define a timeout at all?

salvatore-campagna replied:
As discussed offline I made it as a rest api request param.

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollupShardTask.java:
```diff
@@ -9,15 +9,18 @@
 
-public class RollupShardTask extends CancellableTask {
+public class RollupShardTask extends AllocatedPersistentTask {
+    public static final String TASK_NAME = "rollup-shard";
```
Can we rename this to `downsampel-shard`?

salvatore-campagna replied:
Everything in the code is called "Rollup"<something>...can we, maybe, do this in a separate PR were we rename everything? If I do it here this PR will grow very large.

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollupShardPersistentTaskState.java:
```diff
@@ -0,0 +1,127 @@
+ * @param rollupShardIndexerStatus An instance of {@link RollupShardIndexerStatus} with the rollupShardIndexerStatus of the rollup task
+ * @param tsid The latest successfully processed tsid component of a tuple (tsid, timestamp)
+ */
+public record RollupShardPersistentTaskState(RollupShardIndexerStatus rollupShardIndexerStatus, BytesRef tsid)
```
Maybe also add unit test that extends from `AbstractXContentSerializingTestCase`?

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/TransportDownsampleAction.java:
```diff
@@ -286,143 +297,205 @@ protected void masterOperation(
         }, listener::onFailure));
     }
 
+    private void waitForExistingTask(
```
Maybe we should return a different response here?
Downsampling in progress with a task id that can be used in get task api to see the progress?
If no we should just delegate to `listener` on line 462.

salvatore-campagna committed:
fix: downsampling task disruption

salvatore-campagna committed:
fix: remove unused constant

salvatore-campagna commented on x-pack/plugin/rollup/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/DownsampleClusterDisruptionIT.java:
```diff
@@ -0,0 +1,231 @@
+                    final String restartingNode = randomFrom(dataNodes);
+                    logger.info("Restarting random data node [" + restartingNode + "]");
+                    try {
+                        internalCluster().restartNode(restartingNode);
```
This makes the test fail if the node is not connected...because for instance it was restarted before and still joining?

salvatore-campagna committed:
fix: exceptions thrown and messages

salvatore-campagna committed:
Merge branch 'main' into feature/93582-downsample-persistent-task

salvatore-campagna committed:
test: just restart the downsampling node

salvatore-campagna committed:
fix: use RollupShardIndexerException

salvatore-campagna committed:
fix: remove timeout

salvatore-campagna committed:
fix: timeout as rest param

salvatore-campagna committed:
fix: import DEFAULT_TIMEOUT

salvatore-campagna committed:
Merge branch 'main' into feature/93582-downsample-persistent-task

salvatore-campagna committed:
test: improve testing

salvatore-campagna committed:
fix: inline return statement

salvatore-campagna committed:
fix: remove todo

salvatore-campagna committed:
fix: log executor failures

salvatore-campagna committed:
fix: delagate failure to listener

salvatore-campagna committed:
test: RollupShardIndexer resume capability

salvatore-campagna committed:
fix: remove comments

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/bwc please

salvatore-campagna committed:
fix: remove timeout and fix hashcode and eqauls

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleStep.java:
```diff
@@ -128,7 +131,7 @@ public void performAction(
 
     void performDownsampleIndex(String indexName, String downsampleIndexName, ActionListener<Void> listener) {
```
Thanks for working on this Salvatore

I'm wondering, above at line `104` https://github.com/elastic/elasticsearch/pull/97557/files#diff-54833e304261a22efe3c0a185c58d4742391bd45ef654f1a0b729bcf3374a0e4R107
we seem to continue to fail the downsample operation and try again - the downsample index name 
will remain the same IIUC. Will this not continuously fail as the target downsample index exists already? 

Should we not check with the persistent task framework (or cluster state) if the operation already in progress?

martijnvg replied:
I think we should allow the transport downsample action to be invoked multiple times? If the target index already exists then we can skip that. If the persistent task already exists then we just need to wait for completion of the shard level rollup persistent tasks. 

The downsample action does perform a few post downsample operations like updating index settings, this happens via the elected master node. If ILM retries execution, then I think we should assume that the thread invoking downsample action failed too. So it should be allowed to execute downsample action multiple times.

andreidan replied:
>I think we should allow the transport downsample action to be invoked multiple times? 

Oh sure that sounds good too. Is the TransportAction now treating a downsampling request for a target index that exists already as a no-op?

We have in `main` a race condition where we delete the downsample target index but RollupShardTask is still running and recreates the target index. So in ILM we rewind, delete the target index, generate a new one and try again. 

With this PR the target index name remains stable - is the whole rewind, delete, retry still needed?

andreidan replied:
Coming back to some code above (not able to comment in this PR) 

But this bit  https://github.com/elastic/elasticsearch/pull/97557/files#diff-54833e304261a22efe3c0a185c58d4742391bd45ef654f1a0b729bcf3374a0e4R107 

I don't think we should fail the `DownsampleStep` (as ILM will retry it) if we detect that the target index exists and the downsampling is not complete (as the rollup task can still be running) - note this can also happen if the downsample operation timeouts on the client side (ILM side), not only on master failover.

I think this branch of failure should disappear completely in ILM and just call the downsample API again ? Then the transport downsampling action can identify that the rollup task is running and execute gracefully (i.e. notify the client, ILM, when downsampling completes). 

Also, just a quick question to consider - do we need to identify a complete failure of the rollup task vs a cancelled rollup task vs a time out in ILM and react differently to each of these? 
Ideally ILM should just retry downsampling and the transport downsample action should know how to treat a new incoming downsample request in each of these situations

martijnvg replied:
> Also, just a quick question to consider - do we need to identify a complete failure of the rollup task vs a cancelled rollup task vs a time out in ILM and react differently to each of these?
Ideally ILM should just retry downsampling and the transport downsample action should know how to treat a new incoming downsample request in each of these situations

A cancel will result in error being returned from the downsample action. ILM can then retry and downsample will start from where it left off. (the target index will still exist, only the persistent tasks need to be created. The downsample action will make sure that this happens)

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/downsample/DownsampleAction.java:
```diff
@@ -39,11 +43,18 @@ private DownsampleAction() {
     public static class Request extends MasterNodeRequest<Request> implements IndicesRequest, ToXContentObject {
         private String sourceIndex;
         private String targetIndex;
+        private TimeValue timeout;
```
Maybe rename to wait timeout? To highlight that this is the timeout for waiting for the persistent task to complete?

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/downsample/DownsampleAction.java:
```diff
@@ -25,12 +27,14 @@
 public class DownsampleAction extends ActionType<AcknowledgedResponse> {
     public static final DownsampleAction INSTANCE = new DownsampleAction();
     public static final String NAME = "indices:admin/xpack/downsample";
+    public static final TimeValue DEFAULT_TIMEOUT = new TimeValue(1, TimeUnit.DAYS);
```
Default wait timeout?

martijnvg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleAction.java:
```diff
@@ -149,10 +178,11 @@ public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
         );
 
         // Generate a unique downsample index name and store it in the ILM execution state
```
maybe adjust this comment? Since the index name is no longer unique?

martijnvg commented on x-pack/plugin/rollup/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/DownsampleClusterDisruptionIT.java:
```diff
@@ -0,0 +1,442 @@
+        assertAcked(indicesAdmin().prepareCreate(sourceIndex).setSettings(settings.build()).setMapping(mapping).get());
+    }
+
+    public void testRollupIndexWithDataNodeRestart() throws IOException, InterruptedException {
```
Should we also write such a test where downsampling is triggered from ILM?

andreidan replied:
Awesome we have these disruption tests @salvatore-campagna ❤️ 

++ on Martin's suggestion

I'm curious how the downsample action behaves when it is retried (now that the index name is stable)
Note that in ILM we'd also be retrying when the `downsample` _request_ timeouts

salvatore-campagna committed:
fix: refactor callback hell in TransportDownsampleAction

salvatore-campagna committed:
fix: wait for task completion if it already exists

salvatore-campagna committed:
Merge branch 'main' into feature/93582-downsample-persistent-task

salvatore-campagna committed:
fix: address some review comments

salvatore-campagna committed:
fix: revert changes made by mistake

andreidan commented on x-pack/plugin/rollup/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/DownsampleClusterDisruptionIT.java:
```diff
@@ -0,0 +1,442 @@
+        }
+    }
+
+    private boolean startRollupTask(
```
Would it make sense to document this method? 
It's not immediately apparent what this method does and what the boolean it returns represents (i.e. it does more than just starting a task)

Would `maybeStartRollupTaskDuringDisruption` be a more accurate name?

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateDownsampleIndexNameStep.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Objects;
+import java.util.function.BiFunction;
+
+public class GenerateDownsampleIndexNameStep extends GenerateUniqueIndexNameStep {
```
Can you please document this class

andreidan reviewed (commented):
Thanks for working on this Salvatore

salvatore-campagna committed:
fix: waiting for existing persistent task

salvatore-campagna committed:
fix: extract wait method

salvatore-campagna committed:
fix: just pass number of shards instead of AtomicInteger

salvatore-campagna committed:
fix: document some methods and change a method name

salvatore-campagna committed:
fix: error message while starting persistent task

salvatore-campagna commented:
Some tests like "Downsample admin user without document level or field level security settings defined on the source index" are failing because the target index already exists.

martijnvg committed:
use `persistentRollupTaskId` instead of `persistentTask.getId()`

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/TransportDownsampleAction.java:
```diff
@@ -217,8 +224,11 @@ protected void masterOperation(
+        try {
+            MetadataCreateIndexService.validateIndexName(rollupIndexName, state);
+        } catch (ResourceAlreadyExistsException e) {
+            // ignore index already exists
```
If the index exists, I believe we should check the `index.downsample.status` index setting. In case its value is `SUCCESS` we can skip the whole process.

martijnvg replied:
Good point, we can check the downsample status and short-circuit the process of attempting to create a new index and start the persistent tasks.

csoulios reviewed (commented):
Thanks for taking this task Salvatore. It is a good improvement. I have left some comments

martijnvg committed:
Merge remote-tracking branch 'es/main' into feature/93582-downsample-persistent-task

martijnvg committed:
rename timeout to waitTimeout

martijnvg committed:
fix countdown issue and tidy up test

martijnvg committed:
spotless

martijnvg committed:
fix yaml test

martijnvg committed:
fix yaml test again

martijnvg committed:
Change the downsample action to not fail when downsampling to a target index that already has been downsampled or is being downsampled.
In the former case, the downsample action will return an acked response in the later the downsample action will wait for the previous downsample request to complete and then return an acked response.

martijnvg committed:
make cleanupDownsampleIndexStep a noop step

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleAction.java:
```diff
@@ -138,22 +167,16 @@ public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
-        // Generate a unique downsample index name and store it in the ILM execution state
-        GenerateUniqueIndexNameStep generateDownsampleIndexNameStep = new GenerateUniqueIndexNameStep(
+        // Generate a predictable downsample index name and store it in the ILM execution state
+        GenerateUniqueIndexNameStep generateDownsampleIndexNameStep = new GenerateDownsampleIndexNameStep(
```
If the name is stable I don't think we need to store it in the cluster state (and `GenerateDownsampleIndexNameStep` doesn't need to exist anymore). 

`Downsample` step has the fixed interval so could figure out the name of the target downsample index itself I believe?

andreidan reviewed (commented):
Thanks for working on this Martijn.

Love the direction here. I left a couple of comments.

martijnvg committed:
removed rollup word usages

martijnvg committed:
added disruption test with downsampling managed from ilm

martijnvg committed:
Two changes:
* Make generateDownsampleIndexNameStep a noop and determine the target name in downsample step.
* Don't fail the downsample step if the target index already exists

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleStep.java:
```diff
@@ -89,48 +84,26 @@ public void performAction(
-                listener.onFailure(e);
-            }));
         }
+        performDownsampleIndex(indexName, downsampleIndexName, ActionListener.wrap(listener::onResponse, e -> {
```
Thanks for making the changes Martijn. 

I believe this needs to be in an `else` branch or above after [this line](https://github.com/elastic/elasticsearch/pull/97557/files?diff=split&w=0#diff-54833e304261a22efe3c0a185c58d4742391bd45ef654f1a0b729bcf3374a0e4R94) we need a `return;` (on line 95)

andreidan reviewed (commented):
Thanks Martijn, left one comment on the ILM side

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleStep.java:
```diff
@@ -34,21 +35,24 @@ public class DownsampleStep extends AsyncActionStep {
-        DateHistogramInterval fixedInterval
+        final StepKey key,
+        final StepKey nextStepOnSuccess,
+        final StepKey nextStepOnFailure,
```
ah one more thing, I believe the `DownsampleStep` doesn't need this dance to go back to the `clean` step. It can now be just a regular `AsyncActionStep` with one `nextStepKey` and whenever it fails it'll go into the `ERROR` step and be retried automatically later (now, the downsample transport will support retries without an issue)

See https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java#L34 as an example of a regular async step

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleAction.java:
```diff
@@ -138,32 +167,21 @@ public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
         // Mark source index as read-only
         ReadOnlyStep readOnlyStep = new ReadOnlyStep(readOnlyKey, cleanupDownsampleIndexKey, client);
```
From the readonly step we can no jump straight to the `DownsampleStep` denoted by `downsampleKey` (there's no need to go through the no-op steps).

The noop steps are there just so deployments that are in that step (that we now converted to a noop) when the Elasticsearch upgrade was performed resume the ILM execution and complete the downsample action after upgrade. (perhaps a comment stating the above would help)

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleAction.java:
```diff
@@ -138,32 +167,21 @@ public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
-        );
+        // Before the downsample action was retry-able, we used to generate a unique downsample index name and delete the previous index in
+        // case a failure occurred. The downsample action can now retry execution in case of failure and start where it left off.
+        NoopStep cleanupDownsampleIndexStep = new NoopStep(cleanupDownsampleIndexKey, generateDownsampleIndexNameKey);
```
This step can now go straight to `downsampleKey` (no need to go through the no-op `generateDownsampleIndexNameKey`)

martijnvg committed:
Give test more time to complete downsampling in case failures do occur.

martijnvg committed:
Applied ilm feedback.

martijnvg committed:
Merge remote-tracking branch 'es/main' into feature/93582-downsample-persistent-task

martijnvg committed:
checkstyle

martijnvg committed:
fixed version mistake

martijnvg committed:
stop using `LifecycleState.downsampleIndexName()`

martijnvg committed:
rename rollup to downsample in a couple of places

martijnvg committed:
revert using source index name as part of persistent task id, other we can't execute multiple downsample operations for the same source index concurrently.

martijnvg committed:
Merge remote-tracking branch 'es/main' into feature/93582-downsample-persistent-task

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleStep.java:
```diff
@@ -93,11 +86,9 @@ public void performAction(
                 );
                 listener.onResponse(null);
             }
```
If the index doesn't have the SUCCESS tag but we're re-executing the downsample ILM step we should call `performDownsampleIndex(indexName, downsampleIndexName, ActionListener.wrap(listener::onResponse, listener::onFailure));` to start waiting for a result or get an error message from the transport action

andreidan reviewed (approved):
Thanks for iterating on this Martijn, the ILM side LGTM

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleStep.java:
```diff
@@ -70,16 +67,7 @@ public void performAction(
-            );
-            return;
-        }
+        final String downsampleIndexName = generateDownsampleIndexName(indexName, fixedInterval);
```
This just occurred to me - if we have downsampling configured multiple times, say in hot and warm. The hot phase will generate the target index name `downsample-myindex-1h` and then when we get to the warm phase to execute downsampling again we'll end up with the index name `downsample-downsample-myindex-1h-3h` 
(as we don't use the _original_ index name anymore).

This problem exists already in ILM but perhaps becomes a bit more visible with the new naming convention? 
(currently in ILM we can endup with names like this `downsample-8hit-downsample-eruv-.ds-my-data-2023.08.14-000001` )

Perhaps the Downsampling transport action should configure the index provided name setting `index.provided_name` to maintain the original index name? And we can make use of it in ILM and such to generate the new index name based on the originally provided name?

martijnvg replied:
Right, this will look strange (and already looks strange). However I don't think the downsample action shouldn't be responsible for this, because the contract of this api is that the target index name is provided. 

I think we should make the `DownsampleStep` smarter and that it generates a better name based on the source index? If the source index doesn't contain a the downsample prefix then it generates the same way it does now in the PR. If the source index contains the downsample index name prefix than it parse out the source index and generates a name from that.

I think we should do this in a follow up PR. This PR is already large and the problem already exists today. I will create a follow up PR and label it as a bug. We can then backport this as well after the FF.

martijnvg replied:
Or just use provided index name in DownsampleStep, since current cluster state is available.

andreidan replied:
Great point, thanks for working on this !

martijnvg committed:
Merge remote-tracking branch 'es/main' into feature/93582-downsample-persistent-task

martijnvg committed:
tweaked disruption integration tests.

andreidan commented:
Thanks @salvatore-campagna and @martijnvg for the great work here ! This is great !
